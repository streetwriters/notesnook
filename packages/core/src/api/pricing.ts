/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { SubscriptionPlan, SubscriptionPlanId } from "../types.js";
import hosts from "../utils/constants.js";
import http from "../utils/http.js";

export type SKUResponse = {
  country: string;
  countryCode: string;
  sku?: string;
  discount: number;
};

export interface PlanPrice {
  gross: number;
  net: number;
  tax: number;
  currency?: string;
}

export type Period = "monthly" | "yearly" | "5-year";

export type Discount = {
  type: "regional" | "promo";
  code?: string;
  recurring: boolean;
  amount: number;
};

export interface Plan {
  plan: SubscriptionPlan;
  recurring: boolean;
  id: string;
  name?: string;
  period: Period;
  price: PlanPrice;
  currency: string;
  currencySymbol?: string;
  originalPrice?: PlanPrice;
  discount?: Discount;
  country: string;
}

export class Pricing {
  static sku(
    platform: "google" | "apple" | "paddle" | "paddleNoTrial",
    period: Period,
    plan: SubscriptionPlanId
  ): Promise<SKUResponse> {
    return http.get(
      `${hosts.NOTESNOOK_HOST}/api/v2/prices/skus?platform=${platform}&period=${period}&plan=${plan}`
    );
  }

  static products(trialsAvailed?: SubscriptionPlan[]): Promise<Plan[]> {
    const url = new URL(`${hosts.NOTESNOOK_HOST}/api/v2/prices/products`);
    if (trialsAvailed)
      url.searchParams.set("trialsAvailed", trialsAvailed.join(","));
    return http.get(url.toString());
  }
}
