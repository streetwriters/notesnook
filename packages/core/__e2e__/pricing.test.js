/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Pricing } from "../src/api/pricing.ts";
import { test, expect, describe } from "vitest";

test.each(["monthly", "yearly", undefined])(`get %s price`, async (period) => {
  const price = await Pricing.price(period);
  expect(price).toMatchSnapshot(
    {
      country: expect.any(String),
      countryCode: expect.any(String),
      discount: expect.any(Number),
      price: expect.any(Number)
    },
    `${period || "monthly"}-pricing`
  );
});

describe.each(["android", "ios", "web"])(`get %s pricing tier`, (platform) => {
  test.each(["monthly", "yearly"])(
    `get %s ${platform} tier`,
    async (period) => {
      const price = await Pricing.sku(platform, period);
      expect(price).toMatchSnapshot(
        {
          country: expect.any(String),
          countryCode: expect.any(String),
          discount: expect.any(Number),
          sku: expect.any(String)
        },
        `${period}-${platform}-pricing`
      );
    }
  );
});
