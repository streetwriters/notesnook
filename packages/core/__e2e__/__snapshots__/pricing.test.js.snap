// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get android pricing tier > get monthly android tier > monthly-android-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get android pricing tier > get yearly android tier > yearly-android-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get ios pricing tier > get monthly ios tier > monthly-ios-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get ios pricing tier > get yearly ios tier > yearly-ios-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get monthly price > monthly-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "price": Any<Number>,
}
`;

exports[`get undefined price > monthly-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "price": Any<Number>,
}
`;

exports[`get web pricing tier > get monthly web tier > monthly-web-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get web pricing tier > get yearly web tier > yearly-web-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`get yearly price > yearly-pricing 1`] = `
{
  "country": Any<String>,
  "countryCode": Any<String>,
  "discount": Any<Number>,
  "price": Any<Number>,
}
`;
